### PR TITLE
Flink 1.20: Update Flink to use planned Avro reads

### PR DESCRIPTION
**Pull request by jbonofre**

### CODE REVIEW

**Comment by jbonofre** [comment]
It seems the upsert test is stale. I'm investigating (maybe the schema type mapping).

**Comment by jbonofre** [comment]
The problem seems to be related to:

```
java.lang.ClassCastException: class java.lang.String cannot be cast to class org.apache.flink.table.data.StringData (java.lang.String is in module java.base of loader 'bootstrap'; org.apache.flink.table.data.StringData is in unnamed module of loader 'app')
        at org.apache.flink.table.data.GenericRowData.getString(GenericRowData.java:169)
        at org.apache.flink.table.data.RowData.lambda$createFieldGetter$245ca7d1$1(RowData.java:221)
        at org.apache.iceberg.flink.data.RowDataProjection.getValue(RowDataProjection.java:172)
        at org.apache.iceberg.flink.data.RowDataProjection.isNullAt(RowDataProjection.java:193)
        at org.apache.iceberg.flink.data.RowDataUtil.clone(RowDataUtil.java:96)
        at org.apache.iceberg.flink.source.reader.RowDataRecordFactory.clone(RowDataRecordFactory.java:71)
        at org.apache.iceberg.flink.source.reader.RowDataRecordFactory.clone(RowDataRecordFactory.java:28)
        at org.apache.iceberg.flink.source.reader.ArrayPoolDataIteratorBatcher$ArrayPoolBatchIterator.next(ArrayPoolDataIteratorBatcher.java:98)
        at org.apache.iceberg.flink.source.reader.ArrayPoolDataIteratorBatcher$ArrayPoolBatchIterator.next(ArrayPoolDataIteratorBatcher.java:67)
        at org.apache.iceberg.flink.source.reader.IcebergSourceSplitReader.fetch(IcebergSourceSplitReader.java:96)
        at org.apache.flink.connector.base.source.reader.fetcher.FetchTask.run(FetchTask.java:58)
        at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:165)
        ... 6 more
```

So it seems related to string read from the `GenericRowData`. I'm checking.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkPlannedAvroReader.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+          return ValueReaders.longs();
+        case FLOAT:
+          if (partner != null && partner.typeId() == Type.TypeID.DOUBLE) {
+            return ValueReaders.floatsAsDoubles();
+          }
+          return ValueReaders.floats();
+        case DOUBLE:
+          return ValueReaders.doubles();
+        case STRING:
+          return ValueReaders.strings();
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkPlannedAvroReader.java:179]: Maybe this needs to be `FlinkValueReaders.strings()`?

Comment (jbonofre) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkPlannedAvroReader.java:179]: Yes that's my bad. I'm fixing it.

**Comment by jbonofre** [comment]
I fixed the issue on `ValueReaders` about strings.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroReaderWriter.java [hunk 2]
```diff
@@ -91,7 +91,7 @@ private void writeAndValidate(Schema schema, List<Record> expectedRecords, int n
     try (CloseableIterable<RowData> reader =
         Avro.read(Files.localInput(recordsFile))
             .project(schema)
-            .createReaderFunc(FlinkAvroReader::new)
+            .createResolvingReader(FlinkPlannedAvroReader::create)
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroReaderWriter.java:94]: QQ: Do we have remaining tests for the old reader? I usually try to keep at least a few tests for the deprecated features as well, so they are not broken unintentionally by future changes (and mark them as deprecated, so we don't forget to remove them when the feature is removed).
If there are other tests? Do we test the same functions for the new reader?

**Comment by jbonofre** [comment]
I'm fixing a new `ClassCastException`.

**Comment by jbonofre** [comment]
I fixed all types (primitves and arrays) reads.  Tests should be happy now 😄

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroReaderWriter.java (hunk 2 above)
Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroReaderWriter.java:94]: That's a good point. I can add tests specific to the "old" `FlinkAvroReader`.

**Comment by RussellSpitzer** [comment]
@pvary + @jbonofre how important is this for 1.7? If it is important we need to wrap this up ASAP

**Comment by jbonofre** [comment]
@RussellSpitzer the planned Avro reads has been added to Spark (for Iceberg 1.7.x). This one is not a blocker for 1.7.0 but a good to have to benefit the same performance boost as Spark.

**Review by Fokko — APPROVED** [review]
I agree with @pvary, but apart from that, it looks good to me 👍

**Comment by RussellSpitzer** [comment]
Alright! Chatted with folks and we will do the old-path tests in a follow up.

**Comment by RussellSpitzer** [comment]
Thanks @jbonofre for the PR and @Fokko and @pvary for Review. Let's add some additional tests for Spark and Flink in a followup for the old path.

**Comment by pvary** [comment]
I think we should backport this change to Flink 1.19, and Flink 1.18 as well.

**Pull request merged**